pkg/ruler/ruler.go

```diff
@@ -698,32 +712,35 @@ func (r *Ruler) getShardedRules(ctx context.Context) ([]*GroupStateDesc, error)
 		return nil, fmt.Errorf("unable to inject user ID into grpc request, %v", err)
 	}
 
-	var rgs []*GroupStateDesc
+	var (
+		mergedMx sync.Mutex
+		merged   []*GroupStateDesc
+	)
+
+	// Concurrently fetch rules from all rulers. Since rules are not replicated,
+	// we need all requests to succeed.
+	jobs := concurrency.CreateJobsFromStrings(rulers.GetAddresses())
+	err = concurrency.ForEach(ctx, jobs, len(jobs), func(ctx context.Context, job interface{}) error {
+		addr := job.(string)
 
-	for _, rlr := range rulers.Ingesters {
-		dialOpts, err := r.cfg.ClientTLSConfig.DialOption(nil, nil)
+		grpcClient, err := r.clientsPool.GetClientFor(addr)
 		if err != nil {
-			return nil, err
+			return errors.Wrapf(err, "unable to get client for ruler %s", addr)
 		}
-		conn, err := grpc.DialContext(ctx, rlr.Addr, dialOpts...)
+
+		newGrps, err := grpcClient.(RulerClient).Rules(ctx, &RulesRequest{})
```

Comment:
I believe there's one more bug we need to fix and I think we can test it by having the grpc method `Rules` fail via escape failure.

It seems like gRPC methods don't like when you return `nil` as a pointer to a struct they expect as response because they call `new()` on it e.g. `*RulesResponse` for `Rules(ctx context.Context, in *RulesRequest) (*RulesResponse, error)`

So, we need to do:
```
func (r *Ruler) Rules(ctx context.Context, in *RulesRequest) (*RulesResponse, error) {
	userID, err := tenant.TenantID(ctx)
	if err != nil {
		return &RulesResponse{}, fmt.Errorf("no user id found in context")
	}
	groupDescs, err := r.getLocalRules(userID)
	if err != nil {
		return &RulesResponse{}, err
	}
	return &RulesResponse{Groups: groupDescs}, nil
}
```

Reply:
An example error is:

```
 msg="GET /api/prom/api/v1/rules (500) 4.012183ms Response: \"{\\\"status\\\":\\\"error\\\",\\\"data\\\":null,\\\"errorType\\\":\\\"server_error\\\",\\\"error\\\":\\\"unable to retrieve rules from other rulers, rpc error: code = Internal desc = grpc: error while marshaling: proto: Marshal called with nil\\\"}\"
 ```

Reply:
I don't think the problem is the `return`. We do `return nil, err` in any gRPC function. Why should be different here?

Reply:
> error while marshaling: proto: Marshal called with nil

I've been able to reproduce it on the request side, calling `grpcClient.(RulerClient).Rules(ctx, nil)`  instead of the actual `grpcClient.(RulerClient).Rules(ctx, &RulesRequest{})`. In master we're calling `.Rules(ctx, nil)` and it's something we have broken with the grpc upgrade.

Reply:
See here:
https://github.com/cortexproject/cortex/pull/3917

Reply:
Good catch, I originally thought the problem was intermittent so my hunch was on the return side.